sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -510,32 +542,90 @@ class Analyzer(
+    //   |- view4 (defaultDatabase = db4)
+    // In this case, the view `view1` is a nested view, it directly references `table2`、`view2`
+    // and `view4`, the view `view2` references `view3`. On resolving the table, we look up the
+    // relations `table2`、`view2`、`view4` using the default database `db1`, and look up the
+    // relation `view3` using the default database `db2`.
+    //
+    // Note this is compatible with the views defined by older versions of Spark(before 2.2), which
+    // have empty defaultDatabase and all the relations in viewText have database part defined.
+    def resolveRelation(
+        plan: LogicalPlan,
+        defaultDatabase: Option[String] = None): LogicalPlan = plan match {
+      case u: UnresolvedRelation if !isRunningDirectlyOnFiles(u.tableIdentifier) =>
+        val defaultDatabase = AnalysisContext.get.defaultDatabase
+        val relation = lookupTableFromCatalog(u, defaultDatabase)
+        resolveRelation(relation, defaultDatabase)
+      // The view's child should be a logical plan parsed from the `desc.viewText`, the variable
+      // `viewText` should be defined, or else we throw an error on the generation of the View
+      // operator.
+      case view @ View(desc, _, child) if !child.resolved =>
+        val nestedViewLevel = AnalysisContext.get.nestedViewLevel + 1
+        val context = AnalysisContext(defaultDatabase = desc.viewDefaultDatabase,
+          nestedViewLevel = nestedViewLevel)
+        // Resolve all the UnresolvedRelations and Views in the child.
+        val newChild = AnalysisContext.withAnalysisContext(context) {
+          execute(child)
+        }
+        view.copy(child = newChild)
+      case p @ SubqueryAlias(_, view: View, _) =>
+        val newChild = resolveRelation(view, defaultDatabase)
+        p.copy(child = newChild)
+      case _ => plan
+    }
+
+    def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
+      case i @ InsertIntoTable(u: UnresolvedRelation, parts, child, _, _) if child.resolved =>
+        i.copy(table = EliminateSubqueryAliases(lookupTableFromCatalog(u)))
+      case u: UnresolvedRelation => resolveRelation(u)
+    }
+
+    // Look up the table with the given name from catalog. The database we look up the table from
+    // is decided follow the steps:
+    // 1. If the database part is defined in the table identifier, use that database name;
+    // 2. Else If the defaultDatabase is defined, use the default database name(In this case, no
+    //    temporary objects can be used, and the default database is only used to look up a view);
+    // 3. Else use the currentDb of the SessionCatalog.
+    private def lookupTableFromCatalog(
+        u: UnresolvedRelation,
+        defaultDatabase: Option[String] = None): LogicalPlan = {
       try {
-        catalog.lookupRelation(u.tableIdentifier, u.alias)
+        val tableIdentWithDb = u.tableIdentifier.withDatabase(defaultDatabase)
```

Comment:
Also, it is not really `withDatabase`, right? Because we will not change the database if defaultDatabase is None. 

I feel it is pretty confusing at here. I'd just remove `withDatabase` for now and explicitly check if defaultDatabase is None at here. If defaultDatabase is None, I'd not touch `u.tableIdentifier`.

Also, if this is the part that address the backward compatibility issue, we need to have comment to make it clear.